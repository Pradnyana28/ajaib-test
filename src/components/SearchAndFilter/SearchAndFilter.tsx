import React from 'react';
import { Flex, Input, Text, InputGroup, InputRightAddon, Select } from '@chakra-ui/react';
import { SearchIcon } from '@chakra-ui/icons';

interface SearchAndFilterProps {
    filterValue?: string;

    getSearchValue?: (value: string) => void;
    onFilterChange?: React.ChangeEventHandler<HTMLSelectElement>;
};

const SearchAndFilter = (props: SearchAndFilterProps) => {
    const [searchValue, setSearchValue] = React.useState('');

    const _onSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        setSearchValue(e.target.value);
        props.getSearchValue && props.getSearchValue(e.target.value);
    }

    return (
        <Flex gap={2}>
            <div>
                <Text fontSize='sm'>Search</Text>
                <InputGroup size='sm'>
                    <Input
                        data-testid='search-input'
                        placeholder='Search Keywords'
                        value={searchValue}
                        onChange={_onSearchChange}
                    />
                    <InputRightAddon children={<SearchIcon />} />
                </InputGroup>
            </div>
            <div>
                <Text fontSize='sm'>Gender</Text>
                <Select
                    data-testid='gender-filter'
                    size='sm'
                    defaultValue='all'
                    value={props.filterValue}
                    placeholder='Select Gender'
                    onChange={props.onFilterChange}
                >
                    <option value={'all'} data-testid='gender-filter-all'>All</option>
                    <option value={'male'} data-testid='gender-filter-male'>Male</option>
                    <option value={'female'} data-testid='gender-filter-female'>Female</option>
                </Select>
            </div>
        </Flex>
    );
}

export default SearchAndFilter;