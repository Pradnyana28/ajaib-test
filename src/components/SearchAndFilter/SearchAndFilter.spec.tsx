import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import SearchAndFilter from './SearchAndFilter';
import userEvent from "@testing-library/user-event";

describe('<SearchAndFilter />', () => {
    it('renders without crashing', async () => {
        const { container } = render(<SearchAndFilter />);
        expect(container).toMatchSnapshot();
    });

    it('should have default value for filter component', async () => {
        const { getByTestId } = render(<SearchAndFilter />);
        expect((getByTestId('gender-filter-all') as HTMLOptionElement).selected).toBe(true);
        expect((getByTestId('gender-filter-male') as HTMLOptionElement).selected).toBe(false);
        expect((getByTestId('gender-filter-female') as HTMLOptionElement).selected).toBe(false);
    });

    it('should able to search and change value', async () => {
        const { getByTestId } = render(<SearchAndFilter />) as any;

        fireEvent.change(getByTestId('search-input'), {
            target: {
                value: 'Hello'
            }
        });

        expect(getByTestId('search-input').value).toBe('Hello');
    });

    it('should able to search and change value', async () => {
        const { getByTestId } = render(<SearchAndFilter />);
        const inputElement = getByTestId('search-input') as HTMLInputElement;

        await waitFor(
            () => userEvent.type(inputElement, 'Hello world')
        );

        expect(inputElement.value).toBe('Hello world');
    });
});