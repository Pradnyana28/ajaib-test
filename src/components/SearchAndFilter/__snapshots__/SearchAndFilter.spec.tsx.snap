// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchAndFilter /> renders without crashing 1`] = `
<div>
  <div
    class="css-147dy5z"
  >
    <div>
      <p
        class="chakra-text css-g5hezf"
      >
        Search
      </p>
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          class="chakra-input css-tsqr3a"
          data-testid="search-input"
          placeholder="Search Keywords"
          value=""
        />
        <div
          class="chakra-input__right-addon css-nxc9bd"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <p
        class="chakra-text css-g5hezf"
      >
        Gender
      </p>
      <div
        class="chakra-select__wrapper css-42b2qy"
      >
        <select
          class="chakra-select css-1ik61og"
          data-testid="gender-filter"
        >
          <option
            value=""
          >
            Select Gender
          </option>
          <option
            data-testid="gender-filter-all"
            selected=""
            value="all"
          >
            All
          </option>
          <option
            data-testid="gender-filter-male"
            value="male"
          >
            Male
          </option>
          <option
            data-testid="gender-filter-female"
            value="female"
          >
            Female
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-162mkon"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
